dc/s2n-quic-dc/src/uds/sender.rs

```diff
@@ -66,3 +57,60 @@ impl Sender {
         Ok(())
     }
 }
+
+pub struct SendMsg {
+    sender: Sender,
+    packet: Vec<u8>,
+    fd_to_send: OwnedFd,
+    send_future: Option<Pin<Box<dyn Future<Output = std::io::Result<()>> + Send + Sync>>>,
+}
+
+impl SendMsg {
+    pub fn new(sender: Sender, packet: &[u8], fd_to_send: OwnedFd) -> SendMsg {
+        SendMsg {
+            sender,
+            packet: packet.to_vec(),
+            fd_to_send,
+            send_future: None,
+        }
+    }
+}
+
+impl Future for SendMsg {
+    type Output = std::io::Result<()>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        if this.send_future.is_none() {
+            match this
+                .sender
+                .try_send_nonblocking(&this.packet, this.fd_to_send.as_fd())
+            {
+                Ok(_) => return Poll::Ready(Ok(())),
+                Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                    // Allocate a future on the heap only if initial send returns WouldBlock
+                    let sender = this.sender.clone();
+                    let packet = this.packet.clone();
+                    let fd = this.fd_to_send.try_clone()?;
```

Comment:
I think rather than cloning these, we'd probably want to make SendMsg an enum and switch it to a Future state, i.e., it would move out of these fields?